service/worker/scanner/data_corruption_workflow.go

```diff
@@ -0,0 +1,196 @@
+		var fixResults []invariant.FixResult
+		err := workflow.ExecuteActivity(activityCtx, ExecutionFixerActivity, fixList).Get(activityCtx, &fixResults)
+		if err != nil {
+			logger.Error("failed to run execution fixer activity", zap.Error(err))
+			return err
+		}
+
+		activityOptions = workflow.ActivityOptions{
+			ScheduleToStartTimeout: time.Minute,
+			StartToCloseTimeout:    time.Minute,
+		}
+		activityCtx = workflow.WithActivityOptions(ctx, activityOptions)
+		err = workflow.ExecuteActivity(activityCtx, EmitResultMetricsActivity, fixResults).Get(activityCtx, nil)
+		if err != nil {
+			logger.Error("failed to run execution fixer activity", zap.Error(err))
+			return err
+		}
+
+		workflow.GetMetricsScope(ctx)
+		if signalCount > maxSignalNumber {
+			return workflow.NewContinueAsNewError(ctx, reconciliation.CheckDataCorruptionWorkflowType)
+		}
+	}
+}
+
+func ExecutionFixerActivity(ctx context.Context, fixList []entity.Execution) ([]invariant.FixResult, error) {
+	var result []invariant.FixResult
+	index := 0
+	if activity.HasHeartbeatDetails(ctx) {
+		activity.GetHeartbeatDetails(ctx, &index)
+	}
+
+	for index < len(fixList) {
+		execution := fixList[index]
+		pr, err := getDefaultDAO(ctx, execution.ShardID)
+		if err != nil {
+			return nil, err
+		}
+		request := fetcher.ExecutionRequest{
+			DomainID:   execution.DomainID,
+			WorkflowID: execution.WorkflowID,
+			RunID:      execution.RunID,
+		}
+		concreteExecution, err := fetcher.ConcreteExecution(ctx, pr, request)
+		if err != nil {
+			return nil, err
+		}
+
+		currentExecutionInvariant := invariant.NewOpenCurrentExecution(pr)
+		fixResult := currentExecutionInvariant.Fix(ctx, concreteExecution)
+		result = append(result, fixResult)
+		historyInvariant := invariant.NewHistoryExists(pr)
+		fixResult = historyInvariant.Fix(ctx, concreteExecution)
+		result = append(result, fixResult)
+		activity.RecordHeartbeat(ctx, index)
+		index++
+	}
+	return result, nil
+}
+
+func getDefaultDAO(
+	ctx context.Context,
+	shardID int,
+) (persistence.Retryer, error) {
+	res, ok := ctx.Value(reconciliation.CheckDataCorruptionWorkflowType).(resource.Resource)
+	if !ok {
+		return nil, fmt.Errorf("cannot find key %v in context", reconciliation.CheckDataCorruptionWorkflowType)
+	}
+
+	execManager, err := res.GetExecutionManager(shardID)
+	if err != nil {
+		return nil, err
+	}
+	pr := persistence.NewPersistenceRetryer(execManager, res.GetHistoryManager(), c.CreatePersistenceRetryPolicy())
+	return pr, nil
+}
+
+func EmitResultMetricsActivity(ctx context.Context, fixResults []invariant.FixResult) error {
+	res, ok := ctx.Value(reconciliation.CheckDataCorruptionWorkflowType).(resource.Resource)
+	if !ok {
+		return fmt.Errorf("cannot find key %v in context", reconciliation.CheckDataCorruptionWorkflowType)
+	}
+	scope := res.GetMetricsClient().Scope(metrics.CheckDataCorruptionWorkflowScope)
+	for _, result := range fixResults {
+		scope.IncCounter(metrics.DataCorruptionWorkflowCount)
```

Comment:
are we able to tell the fix result is for which invariant here (or maybe we can emit metric in the same activity as the fix)? I guess might be useful to understand what kind of corruption we are dealing with.

Reply:
Added a metrics tag here.